source/posix/socket.c

```diff
@@ -1367,29 +1385,62 @@ int aws_socket_shutdown_dir(struct aws_socket *socket, enum aws_channel_directio
     return AWS_OP_SUCCESS;
 }
 
+static void s_written_task(struct aws_task *task, void *arg, enum aws_task_status status) {
+    (void)task;
+    (void)status;
+
+    /* this is to handle a race condition when a callback kicks off a cleanup, or the user decides
+     * to close the socket based on something they read (SSL validation failed for example).
+     * if clean_up happens when internal_refcount > 0, socket_impl is kept dangling */
+    struct aws_socket *socket = arg;
+    struct aws_allocator *allocator = socket->allocator;
+    struct posix_socket *socket_impl = socket->impl;
+    ++socket_impl->internal_refcount;
+    socket_impl->written_task_scheduled = false;
+
+    /* Notes about weird loop:
+     * 1) Use `stop_after` node so we only process the initial contents of queue.
```

Comment:
Initial contents of queue when this task is run, ignoring any writes queued during delivery